### PR TITLE
Fix string comparison for AstNodes

### LINKED ISSUE #10829: AQL graph traversal: FILTER p.edge[0].field == "str" is false but included in result (for any "edges" prefix)
## My Environment

* __ArangoDB Version__:        3.6.0
* __Storage Engine__:            RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Kubernetes
* __Configuration__:               default
* __Infrastructure__:               AWS
* __Operating System__:        default as given by kube-arangodb
* __Total RAM in your machine__:        4GB
* __Disks in use__:        
* __Used Package__:              Docker: arangodb/arangodb-preview:3.6.0

## Component, Query & Data

__Affected feature__:
AQL query using web interface

__AQL query (if applicable)__:
```
WITH dog, address, location
FOR person in person
    FOR v,e,p in 1..1 OUTBOUND
        person links
        FILTER (p.edge[0].field == "dogs" and p.vertices[1].age > 0)
            OR p.edge[0].field == "home"
        return v
```

__AQL explain (if applicable)__:
```
Query String (228 chars, cacheable: true):
 WITH dog, address, location
 FOR person in person
     FOR v,e,p in 1..1 OUTBOUND
         person links
         FILTER (p.edge[0].field == "dogs" and p.vertices[1].age > 0)
             OR p.edge[0].field == "home"
         return v
 

Execution plan:
 Id   NodeType                  Site  Est.   Comment
  1   SingletonNode             DBS      1   * ROOT
  2   EnumerateCollectionNode   DBS     10     - FOR person IN person   /* full collection scan, 1 shard(s) */
  9   RemoteNode                COOR    10       - REMOTE
 10   GatherNode                COOR    10       - GATHER 
  3   TraversalNode             COOR     3       - FOR v  /* vertex */, p  /* paths */ IN 1..1  /* min..maxPathDepth */ OUTBOUND person /* startnode */  links
  4   CalculationNode           COOR     3         - LET #6 = (((p.`edge`[0].`field` == "dogs") && (p.`vertices`[1].`age` > 0)) || (p.`edge`[0].`field` == "home"))   /* simple expression */
  5   FilterNode                COOR     3         - FILTER #6
  6   ReturnNode                COOR     3         - RETURN v

Indexes used:
 By   Name   Type   Collection   Unique   Sparse   Selectivity   Fields        Ranges
  3   edge   edge   links        false    false        26.67 %   [ `_from` ]   base OUTBOUND

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections  Options                                  Filter / Prune Conditions
 3   1..1                       links             uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   optimize-traversals
  2   scatter-in-cluster
  3   remove-unnecessary-remote-scatter
```

__Dataset__:
Small test dataset (couple of 10 entries for dog, address, person, and location) - can be shared if needed.

__Size of your Dataset on disk__:

## Steps to reproduce

1. run query

__Problem__:

Running the above query on our sample data set, returns an empty list. It appears, that having a FILTER within a graph traversal containing AND and OR constraints, returns an empty list.

If we run the following query of just having the first constraint
```
WITH dog, address, location
FOR person in person
    FOR v,e,p in 1..1 OUTBOUND
        person links
        FILTER (p.edge[0].field == "dogs" and p.vertices[1].age > 0)
        return v
```
a result set of 15 elements is returned (the results can be provided if needed).

If we separately run the following query of the second constraint
```
WITH dog, address, location
FOR person in person
    FOR v,e,p in 1..1 OUTBOUND
        person links
        FILTER p.edge[0].field == "home"
        return v
```
a result set of 7 elements is returned (the results can be provided if needed).

Running the full query
```
WITH dog, address, location
FOR person in person
    FOR v,e,p in 1..1 OUTBOUND
        person links
        FILTER (p.edge[0].field == "dogs" and p.vertices[1].age > 0)
            OR p.edge[0].field == "home"
        return v
```
however, should return a union of the two contstraints, however only
```
[]
```
is returned.

Replacing the AND operator with an OR however returns results:
```
WITH dog, address, location
FOR person in person
    FOR v,e,p in 1..1 OUTBOUND
        person links
        FILTER (p.edge[0].field == "dogs" or p.vertices[1].age > 0)
            OR p.edge[0].field == "home"
        return v
```
(NOTE: This does not yield the desired result of the initial query but much more, due to `age` being present in persons and not only dogs)

__Expected result__:

A union of the two constraints in between the OR statement (i.e. 15 records + 7 records = 22 records)

I am happy to provide any output, dataset or the Debug Package if needed.

### PR DESCRIPTION
**Pull request by markuspf**
The `stringEquals` method for `AstNode`s with an option for case sensitivity uses `strncmp` (or `strncasecmp` respectively, but using `getStringLength()` for determining the `n` that is compared.

This leads to prefixes of `other` being matched for `stringEquals`.

This is a preliminary fix while I consider using a more C++y solution.

Fixes #10829

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,6 +1,10 @@
 devel
 -----
 
+* Fix string comparison bug that lead to traversal queries accepting
+  prefixes of "edges" and "vertices" to be used as object accessors
+  for p
```
Comment (jsteemann) [inline CHANGELOG]: what is `p` here?

File: arangod/Aql/AstNode.h [hunk 2]
```diff
@@ -524,6 +524,10 @@ struct AstNode {
   /// @brief whether or not a string is equal to another
   bool stringEquals(char const* other, bool caseInsensitive) const;
```
Comment (jsteemann) [inline arangod/Aql/AstNode.h]: this should be removed now

File: arangod/Aql/AstNode.cpp [hunk 3]
```diff
@@ -2838,14 +2838,20 @@ void AstNode::setStringValue(char const* v, size_t length) {
...
 
-bool AstNode::stringEquals(char const* other, bool caseInsensitive) const {
-  if (caseInsensitive) {
-    return (strncasecmp(getStringValue(), other, getStringLength()) == 0);
-  }
-  return (strncmp(getStringValue(), other, getStringLength()) == 0);
+bool AstNode::stringEqualsCaseInsensitive(std::string const& other) const {
+  // Since we're not sure in how much trouble we are with unicode
+  // strings, we assert here that strings we use are 7-bit ASCII
+  TRI_ASSERT(std::any_of(other.begin(), other.end(),
```
Comment (jsteemann) [inline arangod/Aql/AstNode.cpp]: Shouldn't this be an std::all_of, checking that all chars have their highest bit *not* set?

File: arangod/Aql/AstNode.cpp [hunk 4]
```diff
@@ -2838,14 +2838,20 @@ void AstNode::setStringValue(char const* v, size_t length) {
...
+                         [](const char c) { return c & 0xf0; }));
+  return (other.size() == getStringLength() &&
+          strncasecmp(other.c_str(), getStringValue(), getStringLength()) == 0);
 }
 
 bool AstNode::stringEquals(std::string const& other) const {
+  // Since we're not sure in how much trouble we are with unicode
+  // strings, we assert here that strings we use are 7-bit ASCII
+  TRI_ASSERT(std::any_of(other.begin(), other.end(),
```
Comment (jsteemann) [inline arangod/Aql/AstNode.cpp]: Shouldn't this be an std::all_of, checking that all chars have their highest bit *not* set?

File: arangod/Aql/AstNode.cpp (hunk 3 above)
Comment (markuspf) [reply arangod/Aql/AstNode.cpp]: It needs to be (note the other stupid bug with `0xf0` vs `0x80` in there as well...)

```
  TRI_ASSERT(std::none_of(other.begin(), other.end(),
                          [](const char c) { return c & 0x80; }));
```

Comment (jsteemann) [reply arangod/Aql/AstNode.cpp]: Fine with this solution

Comment (markuspf) [reply arangod/Aql/AstNode.cpp]: (or alternatively `std::all_of` with `return c >= 0`)

**Comment by markuspf** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7987/

**Comment by markuspf** [comment]
Jenkins blue

**Pull request merged**